csharp/src/SeedLang.Runtime/IVisualizer.cs

```diff
@@ -13,12 +13,8 @@
 // limitations under the License.
 
 namespace SeedLang.Runtime {
-  // The interface of all visualizer classes.
-  public interface IVisualizer {
-    // Called when a binary expression is evaluated.
-    void OnBinaryExpression(IValue left, IValue right, IValue result);
-
-    // Called when an eval statement is executed.
-    void OnEvalStatement(IValue value);
+  // The interface of visualizers to visualize the given event.
+  public interface IVisualizer<in Event> {
```

Comment:
I tried to remove the contravariance modifier "in" from this line - all the unit tests still passed. Are there any special cases that depend on this contravariance generic parameter? If so, it'd be good to add a unit test to cover that.

The other case is on the contrary side: the contravariance modifier "in" used by `IPublisher<in Visualizer>` is really required, otherwise the unit tests will fail. (It is required by `p is IPublisher<Visualizer> publisher` I think).

Reply:
I will remove the modifier "in" here, because I don't plan to make event classes polymorphic. So the Event parameter doesn't need to be contravariance.
Yes, the contravariance modifier "in" used by `IPublisher<in Visualizer>` is really required. I will add a comment before `p is Publisher<Visualizer> publisher` check to make it clear in the future.